### PR TITLE
Update finch to 0.17, publish async APIs

### PR DESCRIPTION
**Pull request by dlindenkreuz**
[Finch 0.17.0 was just released](https://github.com/sneako/finch/commit/603058bf67b83b4974fb516387cacff47312dcce)! 

This unblocks `Req.async_request` and allows us to `{:halt, ...}` from the `:into` callback option of `Req.request()`.

This PR publishes the features made possible by Finch 0.17.0's `stream_while` and removes the related pre-0.17 version checks. I am not entirely familiar with the Finch codebase, so there might be some accidental leftovers even if tests pass.

### CODE REVIEW

**Comment by dlindenkreuz** [comment]
Cross-linking the related upstream issue for reference: https://github.com/sneako/finch/issues/237

File: lib/req.ex [hunk 1]
```diff
@@ -962,14 +961,10 @@ defmodule Req do
     end
   end
 
-  # TODO: Wait for Finch 0.17
-  @doc false
   def async_request(request, options \\ []) do
```
Comment (wojtekmach) [inline lib/req.ex:964]: Could you keep all these new function doc false for now? I need to do a pass on their docs anyway.

File: lib/req/steps.ex [hunk 2]
```diff
@@ -882,28 +882,8 @@ defmodule Req.Steps do
...
-            raise ArgumentError, "returning {:halt, acc} requires Finch 0.17+"
-
-          other ->
-            raise ArgumentError, "expected {:cont, acc}, got: #{inspect(other)}"
-        end
-      end
-
-      Finch.stream(finch_req, finch_name, acc, fun, finch_options)
-    end
+  defp finch_stream_while(finch_req, finch_name, acc, fun, finch_options) do
```
Comment (wojtekmach) [inline lib/req/steps.ex:885]: Could you remove this defp and use Finch.stream_while directly at the callsite?

File: mix.lock [hunk 3]
```diff
@@ -8,7 +8,7 @@
...
+  "finch": {:hex, :finch, "0.17.0", "17d06e1d44d891d20dbd437335eebe844e2426a0cd7e3a3e220b461127c73f70", [:mix], [{:castore, "~> 0.1 or ~> 1.0", [hex: :castore, repo: "hexpm", optional: false]}, {:mime, "~> 1.0 or ~> 2.0", [hex: :mime, repo: "hexpm", optional: false]}, {:mint, "~> 1.3", [hex: :mint, repo: "hexpm", optional: false]}, {:nimble_options, "~> 0.4 or ~> 1.0", [hex: :nimble_options, repo: "hexpm", optional: false]}, {:nimble_pool, "~> 0.2.6 or ~> 1.0", [hex: :nimble_pool, repo: "hexpm", optional: false]}, {:telemetry, "~> 0.4 or ~> 1.0", [hex: :telemetry, repo: "hexpm", optional: false]}], "hexpm", "8d014a661bb6a437263d4b5abf0bcbd3cf0deb26b1e8596f2a271d22e48934c7"},
```
Comment (wojtekmach) [inline mix.lock:11]: Could you change Finch version requirement in mix.exs too?

**Comment by wojtekmach** [comment]
There is one other Finch change, deprecating :protocol in favor of :protocols. Finch defaults to just :http1 but Req can default to [:http1, :http2]. Lmk if you’d like to look into that but no pressure or rush of course!

**Comment by dlindenkreuz** [comment]
Took a shot at deprecating `:protocol` in favor of `:protocols`. I added a runtime deprecation warning and renamed the field in the docs as well, but this likely needs to be pointed out more explicitly.

**Comment by dlindenkreuz** [comment]
Feel free to adjust to your liking!

**Comment by dlindenkreuz** [comment]
Just saw that `:into` docs should be updated as well to mention that `{:halt, …}` can be returned. Will do when I get back to my desk.

**Comment by wojtekmach** [comment]
Thank you!

**Comment by rubas** [comment]
@wojtekmach 

I would appreciate it, if you could create a new release with this fix as we are getting spammed with warnings.


> There is one other Finch change, deprecating :protocol in favor of :protocols. Finch defaults to just :http1 but Req can default to [:http1, :http2]. Lmk if you’d like to look into that but no pressure or rush of course!

```
warning: :protocol option is deprecated. Use :protocols instead.
  (nimble_options 1.1.0) lib/nimble_options.ex:572: NimbleOptions.validate_value/3
  (nimble_options 1.1.0) lib/nimble_options.ex:558: NimbleOptions.validate_option/3
  (nimble_options 1.1.0) lib/nimble_options.ex:540: NimbleOptions.reduce_options/2
  (elixir 1.16.1) lib/enum.ex:4842: Enumerable.List.reduce/3
  (elixir 1.16.1) lib/enum.ex:2582: Enum.reduce_while/3
  (nimble_options 1.1.0) lib/nimble_options.ex:533: NimbleOptions.validate_options/2
  (nimble_options 1.1.0) lib/nimble_options.ex:508: NimbleOptions.validate_options_with_schema_and_path/3
  (finch 0.18.0) lib/finch.ex:232: Finch.cast_pool_opts/1
  (finch 0.18.0) lib/finch.ex:201: anonymous fn/2 in Finch.pool_options!/1
  (stdlib 5.2) maps.erl:416: :maps.fold_1/4
  (finch 0.18.0) lib/finch.ex:160: Finch.start_link/1
```

**Comment by dlindenkreuz** [comment]
Looks like this was released [in 0.4.9](https://github.com/wojtekmach/req/blob/main/CHANGELOG.md#v049-2024-02-14)!

**Pull request merged**